packages/compose/src/hooks/use-event/index.ts

```diff
@@ -0,0 +1,43 @@
+/**
+ * WordPress dependencies
+ */
+import { useRef, useInsertionEffect, useCallback } from '@wordpress/element';
+
+/**
+ * Any function.
+ */
+export type AnyFunction = ( ...args: any ) => any;
+
+/**
+ * Creates a stable callback function that has access to the latest state and
+ * can be used within event handlers and effect callbacks. Throws when used in
+ * the render phase.
+ *
+ * @param callback The callback function to wrap.
+ *
+ * @example
+ *
+ * ```tsx
+ * function Component(props) {
+ *   const onClick = useEvent(props.onClick);
+ *   React.useEffect(() => {}, [onClick]);
+ * }
+ * ```
+ */
+export default function useEvent< T extends AnyFunction >(
```

Comment:
Close as it is to the official polyfill, this could use some basic tests.

Reply:
Not sure how much value there is in tests here. Can't think of any tests that wouldn't be just testing that the underlying primitives work as expected (useRef, useCallback, useEffect). Also, this is a well-known pattern that's very simple, so it doesn't seems like there's much value either in terms of resilience to future changes, because it's almost guaranteed that there won't be any.

Reply:
Tests are often even better than examples in docs, in terms of understanding the usage and how the hook works. But totally up to you, I'm fine to leave without tests if you feel they'll be useless.

Reply:
Ugh, had a response written out and thought it was submitted. 😧 

---

I'm not opposed to tests at all. Just can't think of a useful one. As I said, all tests I can think of would be directly testing the underlying React primitives (which are already extensively tested and proven), like whether refs or callbacks with no deps (useCallback with empty dep array) are stable references or whether effects are executed in order.

But maybe I'm missing the point. Feel free to suggest a test, I personally think they're not necessary not helpful in this instance.

Reply:
Basic tests for this one would basically verify that:

- `useEvent` returns a stable callback reference across renders/rerenders.
- `useEvent` throws an error if the callback is called during rendering.
- The returned callback works correctly and updates when callback changes.

Again, no strong feelings, leaving it to your judgment as to whether you want to add them.